[SPARK-36137][SQL] HiveShim should fallback to getAllPartitionsOf even if directSQL is enabled in remote HMS

sunchao opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Change `HiveShim.getPartitionsByFilter` to always fallback to use `getAllPartitionsMethod` even if `hive.metastore.try.direct.sql` is set to true in the remote HMS.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

At the moment `getPartitionsByFilter` in `HiveShim` only fallback to use `getAllPartitionsMethod` when `hive.metastore.try.direct.sql` is disabled in the remote HMS, and will fail the query otherwise. However, in certain cases the remote HMS will fallback to use ORM (which only support string type for partition columns) to query the underlying RDBMS **even if this config is set to true**. In this scenario, currently Spark will not be able to recover from the exception and will just fail the query.

For instance, we encountered this bug [HIVE-21497](https://issues.apache.org/jira/browse/HIVE-21497) in HMS running Hive 3.1.2, and Spark was not able to pushdown filter for date column.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

Yes, now if Spark is querying partitions from a remote HMS which throws exception even if `hive.metastore.try.direct.sql` is set to true, Spark will fallback to list all partitions and do the pruning on client side, instead of failing the query.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Tested locally with a HMS instance running 3.1.2. It's pretty hard to add a unit test for this since we don't have a mock HMS.

sunchao committed:
initial commit

sunchao commented:
cc @wangyum @cloud-fan @dongjoon-hyun @viirya

SparkQA commented:
**[Test build #141114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141114/testReport)** for PR 33382 at commit [`b56d888`](https://github.com/apache/spark/commit/b56d8887d534dc6b8758ee25d5f277dcf0cf8fca).

dongjoon-hyun reviewed (commented):
Could you share us your opinion about the performance regressions for the non-failed queries, @sunchao ?

viirya commented:
> At the moment getPartitionsByFilter in HiveShim only fallback to use getAllPartitionsMethod when hive.metastore.try.direct.sql is enabled

Do you mean "when hive.metastore.try.direct.sql is disabled"?

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -876,29 +876,21 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
         try {
           // Hive may throw an exception when calling this method in some circumstances, such as
           // when filtering on a non-string partition column when the hive config key
           // hive.metastore.try.direct.sql is false
```
Update this comment?

sunchao replied:
Hmm I think this comment is still accurate right? it explains why we need to catch the exception.

viirya replied:
> Yes, now if Spark is querying partitions from a remote HMS which throws exception even if hive.metastore.try.direct.sql is set to true, Spark will fallback ...

I think the intention of this PR is Hive also throws exception even the config is set to true?

sunchao replied:
Yeah Hive might throw exception even when it is true, which is what we discovered here. The comment, I think, is on the `try ... catch` clause though, and it uses the case to explain why we need to catch the exception here. I can add a sentence such as "It could also throw exception even if `hive.metastore.try.direct.sql` is set to true in certain cases" if that helps to understand the code.

viirya commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -876,29 +876,21 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+              "this problem.", ex)
             // HiveShim clients are expected to handle a superset of the requested partitions
             getAllPartitionsMethod.invoke(hive, table).asInstanceOf[JSet[Partition]]
-          case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] &&
-              tryDirectSql =>
-            throw QueryExecutionErrors.getPartitionMetadataByFilterError(ex)
```
We won't hit this error anymore?

sunchao replied:
Yes, before there are two cases:
1. (this branch) when `tryDirectSql` is true, we'd throw exception and fail the query
2. when `tryDirectSql` is false, we'd fallback to `getAllPartitionsMethod`.
Now we merge the case 1) into 2) and always fallback to `getAllPartitionsMethod`.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45627/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45627/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45627/

sunchao commented:
> > At the moment getPartitionsByFilter in HiveShim only fallback to use getAllPartitionsMethod when hive.metastore.try.direct.sql is enabled
> 
> Do you mean "when hive.metastore.try.direct.sql is disabled"?

Oops my mistake. Updated.

sunchao commented:
> Could you share us your opinion about the performance regressions for the non-failed queries, @sunchao ?

I'm not sure we should call it a regression since it switch the behavior from failing the query (with no workaround at the moment) to letting it succeed. If there are huge number of partitions, the cost for fetching partition metadata from HMS could increase as we are returning all the partitions instead of the filtered ones. However _normally_ this also means there are more data to be processed so the increased planning time could be amortized in the total job execution time.

dongjoon-hyun commented:
If possible, I'd like to make this as a disabled configuration instead of enforcing it. For normal users which has successful queries don't need to bother with this. And, for the failed queries, can we give a directional guide at the error message to enable this? Is it difficult?

SparkQA commented:
**[Test build #141114 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141114/testReport)** for PR 33382 at commit [`b56d888`](https://github.com/apache/spark/commit/b56d8887d534dc6b8758ee25d5f277dcf0cf8fca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141114/

sunchao committed:
add more comments

SparkQA commented:
**[Test build #141174 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141174/testReport)** for PR 33382 at commit [`69514ff`](https://github.com/apache/spark/commit/69514ff98c4331ca72b6d32b0b48b2f8312ca292).

sunchao commented:
> If possible, I'd like to make this as a disabled configuration instead of enforcing it. For normal users which has successful queries don't need to bother with this. And, for the failed queries, can we give a directional guide at the error message to enable this? Is it difficult?

Hmm you mean introduce another configuration in Spark, and only if that turned on, we'd fallback to `getAllPartitionsMethod` when `hive.metastore.try.direct.sql` is true?

```scala
        val tryDirectSql = hive.getMSC.getConfigValue(tryDirectSqlConfVar.varname,
          tryDirectSqlConfVar.defaultBoolVal.toString).toBoolean
        try {
          getPartitionsByFilterMethod.invoke(hive, table, filter)
            .asInstanceOf[JArrayList[Partition]]
        } catch {
          case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] &&
              !tryDirectSql && !enableFallback =>
            getAllPartitionsMethod.invoke(hive, table).asInstanceOf[JSet[Partition]]
          case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] &&
              tryDirectSql || enableFallback =>
            throw QueryExecutionErrors.getPartitionMetadataByFilterError(ex)
        }
```

SparkQA commented:
**[Test build #141174 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141174/testReport)** for PR 33382 at commit [`69514ff`](https://github.com/apache/spark/commit/69514ff98c4331ca72b6d32b0b48b2f8312ca292).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141174/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45686/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45686/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45686/

HyukjinKwon commented:
Yeah, I think we can have a configuration in Spark, that's disabled by default.

sunchao committed:
address comment

sunchao commented:
I feel adding one more config on top of the existing `tryDirectSql` may make it too complex. What if we introduce a new config and use that to decide whether Spark should fallback to call `getAllPartitionsMethod`? and default it to true but with a helpful message to tell users that they can switch it off to fail the query instead if they really want so?

This should only affect those queries in a few scenarios which they used to fail but now can succeed, which IMO is a better outcome. 

For instance:
```scala
        val tryDirectSqlConfVar = HiveConf.ConfVars.METASTORE_TRY_DIRECT_SQL
        val shouldFallback = SQLConf.get.metastorePartitionPruningFallbackOnException
        try {
          getPartitionsByFilterMethod.invoke(hive, table, filter)
            .asInstanceOf[JArrayList[Partition]]
        } catch {
          case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] &&
              shouldFallback =>
            logWarning("Caught Hive MetaException attempting to get partition metadata by " +
              "filter from Hive. Falling back to fetching all partition metadata, which will " +
              "degrade performance. Modifying your Hive metastore configuration to set " +
              s"${tryDirectSqlConfVar.varname} to true (if it is not true already) may resolve " +
              "this problem. Otherwise, you can set " +
              s"${SQLConf.HIVE_METASTORE_PARTITION_PRUNING_FALLBACK_ON_EXCEPTION.key} " +
              " to false and let the query fail instead.", ex)
            getAllPartitionsMethod.invoke(hive, table).asInstanceOf[JSet[Partition]]
        }
```

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -876,29 +876,24 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
-              !tryDirectSql =>
+          case ex: InvocationTargetException if ex.getCause.isInstanceOf[MetaException] =>
             logWarning("Caught Hive MetaException attempting to get partition metadata by " +
               "filter from Hive. Falling back to fetching all partition metadata, which will " +
```
This fallback can lead to a huge performance slowdown, or even hang the application (e.g. listing millions of partitions)

I think we should be very conservative about fallback. Maybe a config to control this is better, instead of looking at the `directSQL` flag.

sunchao replied:
@cloud-fan thanks for your input. What do you think of the proposal [here](https://github.com/apache/spark/pull/33382#issuecomment-883134666)? We can introduce a flag `SQLConf.get.metastorePartitionPruningFallbackOnException` which default to false. As result, the code will no longer depend on the directSQL flag (currently when the flag is turned off on the remote HMS, Spark will always fallback to list all partitions).

cloud-fan replied:
SGTM

sunchao replied:
@cloud-fan updated the PR with this approach. Let me know what you think.

SparkQA commented:
**[Test build #141789 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141789/testReport)** for PR 33382 at commit [`869f66f`](https://github.com/apache/spark/commit/869f66f77ef4dea97411ef67ffe9d3f9f005697b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46302/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46302/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46302/

SparkQA commented:
**[Test build #141789 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141789/testReport)** for PR 33382 at commit [`869f66f`](https://github.com/apache/spark/commit/869f66f77ef4dea97411ef67ffe9d3f9f005697b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141789/

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:
```diff
@@ -876,29 +876,28 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+          // hive.metastore.try.direct.sql is false. In some cases the remote metastore will throw
+          // exceptions even if the config is true, due to various reasons including the
+          // underlying RDBMS, Hive bugs when generating the filter, etc. For this reason we
+          // always fallback to use `Hive.getAllPartitionsOf` here when the exception happens.
```
let's also mention the new conf

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala:
```diff
@@ -110,12 +113,14 @@ class HivePartitionFilteringSuite(version: String)
-    val filteredPartitions = client.getPartitionsByFilter(client.getTable("default", "test"),
-      Seq(attr("ds") === 20170101))
+  test(s"getPartitionsByFilter returns all partitions when $fallbackKey=false") {
+    withSQLConf(SQLConf.HIVE_METASTORE_PARTITION_PRUNING_FALLBACK_ON_EXCEPTION.key -> "true") {
```
nit: we can use `fallbackKey` here

cloud-fan reviewed (approved):
LGTM except for some minor comments

sunchao committed:
address comments

sunchao commented:
Thanks @cloud-fan . Addressed the comments and also added one more test. Also brought back `getPartitionMetadataByFilterError` so users will get a more informative error message.

SparkQA commented:
**[Test build #141960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141960/testReport)** for PR 33382 at commit [`32fc005`](https://github.com/apache/spark/commit/32fc0054a6f71565f9e44d3a542310392ab3bd77).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46471/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46471/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46471/

SparkQA commented:
**[Test build #141960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141960/testReport)** for PR 33382 at commit [`32fc005`](https://github.com/apache/spark/commit/32fc0054a6f71565f9e44d3a542310392ab3bd77).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141960/

viirya commented:
Thanks! Merging to master.

sunchao commented:
Thanks all for the review!

HyukjinKwon reviewed (commented):
LGTM2
